p2p/tests/buckets.rs

```diff
@@ -0,0 +1,176 @@
+    }
+
+    #[test]
+    fn test_different_sk_different_bucket() {
+        let sk_1 = 1;
+        let sk_2 = 2;
+        let src_addr = ip("127.0.0.1:21337");
+        let peer_addr = ip("192.168.1.1:21337");
+
+        let new_index_sk_1 = new_bucket_index(sk_1, &peer_addr, &src_addr);
+        let new_index_sk_2 = new_bucket_index(sk_2, &peer_addr, &src_addr);
+
+        assert_ne!(new_index_sk_1, new_index_sk_2);
+    }
+}
+
+/// Tests for the business logic of inserting peer addresses into the `tried` buckets.
+mod tried {
+    use super::{ip, tried_bucket_index};
+
+    #[test]
+    fn test_same_peer_ip_different_peer_port_same_tried_bucket_different_index() {
+        let sk = 0;
+        let peer_addr_21337 = ip("192.168.1.1:21337");
+        let peer_addr_21338 = ip("192.168.1.1:21338");
+
+        let new_index_21337 = tried_bucket_index(sk, &peer_addr_21337);
+        let new_index_21338 = tried_bucket_index(sk, &peer_addr_21338);
+
+        assert_ne!(new_index_21337, new_index_21338);
+        assert!((f64::from(new_index_21337) - f64::from(new_index_21338)).abs() < 64.0);
+    }
+
+    #[test]
+    fn test_close_peer_ip_same_tried_bucket_different_index() {
+        let sk = 0;
+        let peer_addr_1_1 = ip("192.168.1.1:21337");
+        let peer_addr_1_2 = ip("192.168.1.2:21337");
+
+        let new_index_1_1 = tried_bucket_index(sk, &peer_addr_1_1);
+        let new_index_1_2 = tried_bucket_index(sk, &peer_addr_1_2);
+
+        assert_ne!(new_index_1_1, new_index_1_2);
+        assert!((f64::from(new_index_1_1) - f64::from(new_index_1_2)).abs() < 64.0);
+    }
+
+    #[test]
+    fn test_slightly_far_peer_ip_different_tried_bucket_different_index() {
+        let sk = 0;
+        let peer_addr_1_1 = ip("192.168.1.1:21337");
+        let peer_addr_2_1 = ip("192.168.2.1:21337");
+
+        let new_index_1_1 = tried_bucket_index(sk, &peer_addr_1_1);
+        let new_index_2_1 = tried_bucket_index(sk, &peer_addr_2_1);
+
+        assert_ne!(new_index_1_1, new_index_2_1);
+        assert!((f64::from(new_index_1_1) - f64::from(new_index_2_1)).abs() < 1024.0);
+    }
+
+    #[test]
+    fn test_much_far_peer_ip_different_tried_bucket_different_index() {
+        let sk = 0;
+        let peer_addr_168 = ip("192.168.1.1:21337");
+        let peer_addr_169 = ip("192.169.1.1:21337");
+
+        let new_index_168 = tried_bucket_index(sk, &peer_addr_168);
+        let new_index_169 = tried_bucket_index(sk, &peer_addr_169);
+
+        assert_ne!(new_index_168, new_index_169);
+    }
+
+    #[test]
+    fn test_different_sk_different_bucket() {
+        let sk_1 = 1;
+        let sk_2 = 2;
+        let peer_addr = ip("192.168.1.1:21337");
+
+        let new_index_sk_1 = tried_bucket_index(sk_1, &peer_addr);
+        let new_index_sk_2 = tried_bucket_index(sk_2, &peer_addr);
+
+        assert_ne!(new_index_sk_1, new_index_sk_2);
+    }
+}
+
+fn new_bucket_index(sk: u64, socket_addr: &SocketAddr, src_socket_addr: &SocketAddr) -> u16 {
```

Comment:
This looks like it shouldn't be here, why not use `new_bucket_index` from `p2p` crate?

Reply:
> This looks like it shouldn't be here, why not use `new_bucket_index` from `p2p` crate?

Because the other is a method and takes `&self`.